### DISCUSSION TITLE
Success() does not works

### DISCUSSION THREAD

**Discussion by yi-Xu-0100** [community/community#26898] [Actions]
<p>I use <code>success()</code> in my workflows, but after the previous task was cancelled, success did not cancel the trigger, and caused the error in the subsequent steps?<br>
Is there something wrong with me?</p>
<p>my_workflows.yml</p>
<pre><code class="lang-auto">      - name: setup and build
        if: steps.comparison1.outputs.result == 'true'
        run: |
          npm install
          npm run build-update

      - name: Get menifest version
        if: success()
        uses: actions/github-script@v3.0.0
        id: get-menifest-version
        with:
          result-encoding: string
          script: |
            const manifest = require(`${process.env.GITHUB_WORKSPACE}/build/manifest.json`);
            return manifest.version;
</code></pre>
<p>and the performance:</p>
<p></p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/2X/8/89bd184ba235cbf94a275e7374e387f04b98eaff.png" data-download-href="/uploads/short-url/jEuxWgkGY55FqHPgHPs5oAAe34H.png?dl=1" title="image" rel="noopener nofollow ugc"><img src="https://user-images.githubusercontent.com/25523606/181100517-bf8a5fa5-6e53-4aad-8e90-6e8be2624a93.png" alt="image" data-base62-sha1="jEuxWgkGY55FqHPgHPs5oAAe34H" width="327" height="500" srcset="https://aws1.discourse-cdn.com/github/optimized/2X/8/89bd184ba235cbf94a275e7374e387f04b98eaff_2_327x500.png, https://aws1.discourse-cdn.com/github/original/2X/8/89bd184ba235cbf94a275e7374e387f04b98eaff.png 1.5x, https://aws1.discourse-cdn.com/github/original/2X/8/89bd184ba235cbf94a275e7374e387f04b98eaff.png 2x" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/2X/8/89bd184ba235cbf94a275e7374e387f04b98eaff_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use xlink:href="#far-image"></use></svg><span class="filename">image</span><span class="informations">390×596 22.4 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use xlink:href="#discourse-expand"></use></svg>
</div></a></div>

**Comment by kingthorin** [comment]
<p>Since the last step that reported a status was a success I guess it continued this one instead of skipping.</p>
<p>If it’s really a problem it seems you should check the outcome of each step, or check for <code>cancelled</code>:<br>
</p><aside class="onebox allowlistedgeneric">
  <header class="source">
      <img src="https://docs.github.com/assets/images/site/favicon.svg" class="site-icon" width="16" height="16">
      <a href="https://docs.github.com/en/actions/reference/context-and-expression-syntax-for-github-actions#steps-context" target="_blank" rel="noopener nofollow ugc">docs.github.com</a>
  </header>
  <article class="onebox-body">
    <img src="" class="thumbnail" width="16" height="16">

<h3><a href="https://docs.github.com/en/actions/reference/context-and-expression-syntax-for-github-actions#steps-context" target="_blank" rel="noopener nofollow ugc">Context and expression syntax for GitHub Actions - GitHub Docs</a></h3>



  </article>
  <div class="onebox-metadata">
    
    
  </div>
  <div style="clear: both"></div>
</aside>

**Comment by yi-Xu-0100** [comment]
<p>But in docs:</p>
<blockquote>
<p>success<br>
Returns <code>true</code> when none of the previous steps have failed or been canceled.</p>
</blockquote>
<p><a href="https://docs.github.com/en/actions/reference/context-and-expression-syntax-for-github-actions#success" rel="noopener nofollow ugc">Context and expression syntax for GitHub Actions - GitHub Docs</a></p>

**Comment by kingthorin** [comment]
<p>Could you link the workflow run?</p>

**Comment by yi-Xu-0100** [comment]
<p>I tested the judgment using the <code>outcome</code> and it was successful, but I am still confused about why <code>success()</code> cannot judge success. This is the record at the time.</p>
<p><a href="https://github.com/yi-Xu-0100/soulsign-chrome/runs/1120614823?check_suite_focus=true" rel="noopener nofollow ugc">yi-Xu-0100/soulsign-chrome@09a106b (github.com)</a></p>

**Comment by kingthorin** [comment]
<p>I’m not so sure your cancel was actually processed at all. The “setup and build” step actually says it was skipped, which I believe means that the condition <code>steps.comparison1.outputs.result == 'true'</code> was not met.</p>
<p>Subsequent steps were skipped cause success() wasn’t true after “Get menifest version” failed. (“Get menufest version” failed because success was still true and the file it needed wasn’t present. It doesn’t look like there’s a way to check skipped steps … at least not as a success/failure/cancelled condition [that I’ve seen]).</p>
<p>Note, not sure if you care but there’s a few occurrences of "menifest’ which should likely be “manifest” :wink:</p>

**Comment by yi-Xu-0100** [comment]
<p>Thank you for your reply. You mean that the success status judged by <code>Get manifest version</code> comes from <code>Compare version</code>, and <code>setup and build</code> is skipped directly?</p>
<p>I think it may be caused by such a situation. Is there any judgment of skip state? I only saw the ones that were judged to be cancelled. To determine the state of the previous step, you need to add an id to each state and write a long judgment formula, which is too verbose.</p>
<p>Finally, thank you for pointing out my spelling mistakes, I am always so careless and I will fix this. :smiley:</p>

**Comment by kingthorin** [comment, marked as answer]
<aside class="quote no-group" data-username="yi-Xu-0100" data-post="7" data-topic="132463">
<div class="title">
<div class="quote-controls"></div>
 yi-Xu-0100:</div>
<blockquote>
<p>Thank you for your reply. You mean that the success status judged by <code>Get manifest version</code> comes from <code>Compare version</code> , and <code>setup and build</code> is skipped directly?</p>
</blockquote>
</aside>
<p>Yes I think that’s what’s going on.</p>
<aside class="quote no-group">
<blockquote>
<p>I think it may be caused by such a situation. Is there any judgment of skip state? I only saw the ones that were judged to be cancelled. To determine the state of the previous step, you need to add an id to each state and write a long judgment formula, which is too verbose.</p>
</blockquote>
</aside>
<p>I just did a bit more quick reading, I think you could do something like:</p>
<pre><code class="lang-auto">      - name: setup and build
        id: build
        if: steps.comparison1.outputs.result == 'true'
        run: |
          npm install
          npm run build-update
      - name: Get menifest version
        if: success() &amp;&amp; steps.build.conclusion != 'skipped'
        uses: actions/github-script@v3.0.0
        id: get-menifest-version
        with:
          result-encoding: string
          script: |
            const manifest = require(`${process.env.GITHUB_WORKSPACE}/build/manifest.json`);
            return manifest.version;
</code></pre>
<p>(^ Add <code>id: build</code> to the setup and build step. Add a secondary condition to the Get manifest version step: <code>&amp;&amp; steps.build.conclusion != 'skipped'</code>.)</p>
<aside class="quote no-group">
<blockquote>
<p>Finally, thank you for pointing out my spelling mistakes, I am always so careless and I will fix this.</p>
</blockquote>
</aside>
<p>No problem, happens to all of us.</p>

**Comment by yi-Xu-0100** [comment]
<p>Because the subsequent steps to publish the release also need the support of the previous steps, I have judged the subsequent related steps as an outcome, and there is no problem with the current tests.<br>
Thank you for your suggestions and help. :smiley:</p>
